---
knit: bookdown::preview_chapter
---

# scater package

## Introduction

[scater](https://github.com/davismcc/scater) is a single-cell analysis toolkit for expression with R. This package contains useful tools for the analysis of single-cell gene expression data using the statistical software R. The package places an emphasis on tools for quality control, visualisation and pre-processing of data before further downstream analysis.

[scater](https://github.com/davismcc/scater) enables the following:

* Transcript quantification from raw read files using Kallisto (see [this preprint](http://arxiv.org/abs/1505.02710) for details)
* Data format standardisation and integration with other Bioconductor packages (eg. DESeq)
* Computation of QC metrics
* Many visualisation methods
* Simple normalisation methods

We will be using [scater](https://github.com/davismcc/scater) to demonstrate QC & normalization of single-cell RNASeq data.


## Terminology
In the context of [scater](https://github.com/davismcc/scater) in this course
* Features correspond to genes.
* Samples correspond to individual cells.

However, the methods are generalizable to any genomic or transcriptomic regions.

## SCESet class

Data format standardization is achieved using the SCESet class which organizes your data into a format [scater](https://github.com/davismcc/scater) can manipulate behind-the-scenes for use in a wide variety of existing tools.

 The class requires three input files:

* __exprs__, a numeric matrix of expression values, where rows are features, and columns are cells
* __phenoData__, a table (technically an _AnnotatedDataFrame_ object), where rows are cells, and columns are cell attributes (such as cell type, culture condition, day captured, etc.)
* __featureData__, a table (technically an  _AnnotatedDataFrame_ object), where rows are features (e.g. genes), and columns are feature attributes, such as biotype, gc content, etc.
