---
# knit: bookdown::preview_chapter
output: html_document
---

# Ordering of cells according to pseudotime

```{r, echo=FALSE}
library(knitr)
opts_chunk$set(fig.align = "center")
```
```{r, echo=TRUE, message=FALSE, warning=FALSE}
library(TSCAN)
library(M3Drop)
library(monocle)
set.seed(1)
```

In many situations, one is studying a process where cells change
continuously. This includes for example many differentiation processes
taking place during development, where following a stimulus, cells
will change from one cell-type to another. Ideally, we would like to
monitor the expression levels of an individual cell over
time. Unfortunately, such monitoring is not possible with scRNA-seq
since the cell is lysed (destroyed) when the RNA is extracted.

Instead, we must sample at multiple time-points and obtain snapshots of
the gene expression profiles. Since some of the cells will
proceed faster along the differentiation than others, each snapshot may contain
cells at varying points along the developmental progression. We use statistical methods
 to order the cells along one or more trajectories which represent the underlying
 developmental trajectories, this ordering is referred to as "pseudotime".

In this chapter we will consider two different tools, Monocle and
TSCAN for ordering cells according to their pseudotime development. To
illustrate the methods we will be using a dataset on mouse embryonic
development [@Deng2014-mx]. The dataset consists of
268 cells from 10 different time-points of early mouse development.

## TSCAN

TSCAN combines clustering with pseudotime analysis. First it clusters the cells using `mclust`, 
which is based on a mixture of normal distributions. Then it builds a minimum spanning tree to connect the clusters together. The branch of this tree that connects the largest number of clusters is the main branch which is used to determine pseudotime.

First we will try to use all genes to order the cells.
```{r tscan-all-genes}
deng <- readRDS("deng/deng.rds")
cellLabels <- colnames(deng)
procdeng <- TSCAN::preprocess(deng)
colnames(procdeng) <- 1:268
dengclust <- TSCAN::exprmclust(procdeng, clusternum = 10)
TSCAN::plotmclust(dengclust)
dengorderTSCAN <- TSCAN::TSCANorder(dengclust, orderonly = F)
pseudotime_order_tscan <- as.character(dengorderTSCAN$sample_name)
```

We can also examine which timepoints have been assigned to each state:

```{r tscan-vs-truth}
cellLabels[dengclust$clusterid == 1]
colours <- rainbow(n = 10) # red = early, violet = late
plot(
    cellLabels[as.numeric(pseudotime_order_tscan)], 
    xlab="Pseudotime Order", 
    ylab="Timepoint",
    col = colours[as.numeric(cellLabels[as.numeric(pseudotime_order_tscan)])],
    pch = 16
)
```

__Exercise 1__ Compare results for different numbers of clusters (`clusternum`).

## monocle

Monocle skips the clustering stage of TSCAN and directly builds a minimum spanning tree to connect all cells. Monocle then identifies the longest path in this tree to determine pseudotime. If the data contains diverging trajectories (i.e. one cell type differentiates into two different cell-types), monocle can identify alternative long paths in the tree using the argument `num_paths`. Each of the resulting forked paths is defined as a separate cell state, thus `num_paths = 2` will identify three different cell states.

Unfortunately, Monocle does not work when all the genes are used, so
we must carry out feature selection. First, we use M3Drop:
```{r m3d-select-genes}
m3dGenes <- as.character(
    M3Drop::M3DropDifferentialExpression(deng, suppress.plot = T)$Gene
)
d <- deng[which(m3dGenes %in% rownames(deng)),]
d <- d[!duplicated(rownames(d)),]
```

Now run monocle:
```{r monocle-all-genes, message=FALSE, warning=FALSE}
pd <- as.data.frame(as.numeric(as.character(cellLabels)))
names(pd) <- "timepoint"
pd <- new("AnnotatedDataFrame", data=pd)
fd <- as.data.frame(rownames(d))
names(fd) <- "gene"
fd <- new("AnnotatedDataFrame", data=fd)
colnames(d) <- 1:dim(d)[2]
rownames(d) <- 1:dim(d)[1]
rownames(pd) <- colnames(d)
rownames(fd) <- rownames(d)
dCellData <- monocle::newCellDataSet(d, phenoData = pd, featureData = fd)
dCellData <- monocle::setOrderingFilter(dCellData, 1:length(m3dGenes))
dCellDataSet <- monocle::reduceDimension(dCellData, pseudo_expr = 1)
dCellDataSet <- monocle::orderCells(dCellDataSet, reverse = F, num_paths = 1)
monocle::plot_spanning_tree(dCellDataSet)
# Store the ordering
pseudotime_monocle <- 
    data.frame(
        Timepoint = phenoData(dCellDataSet)$timepoint, 
        pseudotime = phenoData(dCellDataSet)$Pseudotime, 
        State=phenoData(dCellDataSet)$State
    )
rownames(pseudotime_monocle) <- 1:268
pseudotime_order_monocle <- 
    rownames(pseudotime_monocle[order(pseudotime_monocle$pseudotime), ])
```

We can again compare the inferred pseudotime to the known sampling timepoints.
```{r monocle-vs-truth}
plot(
    pseudotime_monocle$pseudotime, 
    pseudotime_monocle$Timepoint, 
    xlab="Pseudotime", 
    ylab="Timepoint",
    col = colours[pseudotime_monocle$Timepoint],
    pch = 16
)
```

## Comparison of the methods

How do the trajectories inferred by TSCAN and Monocle compare?
```{r tscan-monocle-compare}
matched_ordering <- 
    match(
        pseudotime_order_tscan, 
        pseudotime_order_monocle
    )
timepoint_ordered <- 
    pseudotime_monocle[order(pseudotime_monocle$pseudotime), 1]
plot(
    matched_ordering, 
    xlab = "Monocle Order", 
    ylab = "TSCAN Order", 
    col = colours[timepoint_ordered], 
    pch = 16
)
```

## Expression of genes through time
Each package also enables the visualization of expression through pseudotime.

__TSCAN__
```{r Obox6-tscan}
colnames(deng) <- 1:268
TSCAN::singlegeneplot(
    deng[rownames(deng) == "Obox6", ], 
    dengorderTSCAN
)
```

__Monocle__
```{r Obox6-monocle}
monocle::plot_genes_in_pseudotime(
    dCellDataSet[fData(dCellDataSet)$gene == "Obox6",], 
    color_by = "timepoint"
)
```

__Exercise 2__: Repeat the exercise using a subset of the genes, e.g. the set of highly variable genes that can be obtained using M3Drop::BrenneckeGetVariableGenes
